src/direct.js

```diff
@@ -0,0 +1,468 @@
+    );
+
+    this._sdkMetricsRegistry.newGauge(
+      'histograms.queue.size',
+      this._histogramsBuffer.size
+    );
+    this._sdkMetricsRegistry.newGauge(
+      'histograms.queue.remaining_capacity',
+      this._histogramsBuffer.remainCapacity
+    );
+    this._histogramsValid = this._sdkMetricsRegistry.newCounter(
+      'histograms.valid'
+    );
+    this._histogramsInvalid = this._sdkMetricsRegistry.newCounter(
+      'histograms.invalid'
+    );
+    this._histogramsDropped = this._sdkMetricsRegistry.newCounter(
+      'histograms.dropped'
+    );
+    this._histogramsReportErrors = this._sdkMetricsRegistry.newCounter(
+      'histograms.report.errors'
+    );
+
+    this._sdkMetricsRegistry.newGauge(
+      'spans.queue.size',
+      this._tracingSpansBuffer.size
+    );
+    this._sdkMetricsRegistry.newGauge(
+      'spans.queue.remaining_capacity',
+      this._tracingSpansBuffer.remainCapacity
+    );
+    this._spansValid = this._sdkMetricsRegistry.newCounter('spans.valid');
+    this._spansInvalid = this._sdkMetricsRegistry.newCounter('spans.invalid');
+    this._spansDropped = this._sdkMetricsRegistry.newCounter('spans.dropped');
+    this._spansReportErrors = this._sdkMetricsRegistry.newCounter(
+      'spans.report.errors'
+    );
+
+    this._sdkMetricsRegistry.newGauge(
+      'span_logs.queue.size',
+      this._spanLogsBuffer.size
+    );
+    this._sdkMetricsRegistry.newGauge(
+      'span_logs.queue.remaining_capacity',
+      this._spanLogsBuffer.remainCapacity
+    );
+    this._spanLogsValid = this._sdkMetricsRegistry.newCounter(
+      'span_logs.valid'
+    );
+    this._spanLogsInvalid = this._sdkMetricsRegistry.newCounter(
+      'span_logs.invalid'
+    );
+    this._spanLogsDropped = this._sdkMetricsRegistry.newCounter(
+      'span_logs.dropped'
+    );
+    this._spanLogsReportErrors = this._sdkMetricsRegistry.newCounter(
+      'span_logs.report.errors'
+    );
+  }
+
+  /**
+   * Schedule a timer that flush every 5 seconds by default
+   */
+  _scheduleTimer() {
+    this._timer = setInterval(() => {
+      this._batchReport(
+        this._metricsBuffer.toArray(),
+        constants.WAVEFRONT_METRIC_FORMAT,
+        'points',
+        this._pointsReportErrors
+      );
+      this._batchReport(
+        this._histogramsBuffer.toArray(),
+        constants.WAVEFRONT_HISTOGRAM_FORMAT,
+        'histograms',
+        this._histogramsReportErrors
+      );
+      this._batchReport(
+        this._tracingSpansBuffer.toArray(),
+        constants.WAVEFRONT_TRACING_SPAN_FORMAT,
+        'spans',
+        this._spansReportErrors
+      );
+      this._batchReport(
+        this._spanLogsBuffer.toArray(),
+        constants.WAVEFRONT_SPAN_LOG_FORMAT,
+        'span_logs',
+        this._spanLogsReportErrors
+      );
+    }, this._flushIntervalSeconds);
+  }
+
+  /**
+   * Report data to server, and record metrics
+   */
+  async _reportToServer(options, dataFormat, entityPrefix, reportErrors) {
```

Comment:
How is dataFormat generated? is it user-provided? do we sanitize this value?

Reply:
It's defined in the code when we schedule the timer.